RealmSwift/Query.swift

```diff
@@ -0,0 +1,1075 @@
+        var copy = lhs
+        copy.context.expression.insert(.special(.openParentheses), at: 0)
+        return copy.append(expression: [.compound(.and)] + rhs.context.expression + [.special(.closeParentheses)],
+                           subqueryCount: rhs.context.count)
+    }
+    /// :nodoc:
+    public static func || (_ lhs: Query, _ rhs: Query) -> Query {
+        // Wrap the left expression and right expression in parentheses
+        var copy = lhs
+        copy.context.expression.insert(.special(.openParentheses), at: 0)
+        return copy.append(expression: [.compound(.or)] + rhs.context.expression + [.special(.closeParentheses)],
+                           subqueryCount: rhs.context.count)
+    }
+
+    // MARK: Subscript
+
+    /// :nodoc:
+    public subscript<V>(dynamicMember member: KeyPath<T, V>) -> Query<V> where T: ObjectBase {
+        let name = _name(for: member)
+        return append(expression: [.keyPath(name: name)])
+    }
+    /// :nodoc:
+    public subscript<V: RealmCollectionBase>(dynamicMember member: KeyPath<T, V>) -> Query<V> where T: ObjectBase {
+        let name = _name(for: member)
+        return append(expression: [.keyPath(name: name, isCollection: true)])
+    }
+
+    // MARK: Query Construction
+
+    /// Creates an NSPredicate compatibe string.
+    /// - Parameter isSubquery: States if expression need to be arraged in a special way to cater to subqueries.
+    /// - Returns: A tuple containing the predicate string and an array of arguments.
+    public func _constructPredicate(_ isSubquery: Bool = false) -> QueryPredicate {
+        var predicateString: [String] = []
+        var arguments: [Any] = []
+        var subqueryCount = context.count
+
+        func optionsStr(_ options: Set<StringOptions>?) -> String {
+            guard let o = options, !o.isEmpty else {
+                return ""
+            }
+            var str = "["
+            if o.contains(.caseInsensitive) {
+                str += "c"
+            }
+            if o.contains(.diacriticInsensitive) {
+                str += "d"
+            }
+            str += "]"
+            return str
+        }
+
+        func comparison(_ expression: QueryExpression.Comparision, _ idx: Int) {
+            switch expression {
+            case let .between(low, high, closedRange):
+                if closedRange {
+                    predicateString.append(" BETWEEN {%@, %@}")
+                    arguments.append(contentsOf: [low, high])
+                } else if idx > 0, case let .keyPath(name, _) = context.expression[idx-1] {
+                    predicateString.append(" >= %@")
+                    arguments.append(low)
+                    predicateString.append(" && \(name) <\(closedRange ? "=" : "") %@")
+                    arguments.append(high)
+                } else {
+                    throwRealmException("Could not construct .contains(_:) predicate")
```

Comment:
This error message mentions the wrong thing, but more importantly isn't actionable. If a user writes some code which hits this, the error doesn't tell them what the problem is or give them hints to what they need to change to fix it.